src/bg_cache_worker.ts

```diff
@@ -0,0 +1,151 @@
+import * as schemas from "./schemas";
+import Util from "./lib/util";
+
+const network = Arc.Config.get('network');
+
+if (process.env.NODE_ENV == 'production') {
+  // Use Infura on production
+  const infuraNetwork = network == 'live' ? 'mainnet' : network;
+
+  // The default account in reading from Infura, I'm not totally sure why this is needed (but it is), or what account it should be
+  const mnemonic = process.env.DEFAULT_ACCOUNT_MNEMONIC;
+
+  const provider = new HDWalletProvider(mnemonic, "https://" + infuraNetwork + ".infura.io/UeW8cwaou03qFgsAHoDP");
+
+  // Setup web3 ourselves so we can use Infura instead of letting Arc.js setup web3
+  global.web3 = new Web3(provider.engine);
+
+  process.env.API_URL = process.env.API_URL || "https://daostack-alchemy.herokuapp.com";
+} else {
+  process.env.API_URL = process.env.API_URL || "http://127.0.0.1:3001";
+}
+
+const cacheBlockchain = async () => {
+  // TODO: store last block checked somewhere (redis?) and only load new data since last check
+  let lastBlock = 0;
+
+  // tslint:disable-next-line:no-console
+  console.log("Starting to cache the blockchain from block ", lastBlock);
+
+  let initialState: IArcState = arcInitialState;
+  const daos = {} as { [key: string]: IDaoState };
+
+  const daoCreator = await Arc.DaoCreator.deployed();
+
+  try {
+    // Get the list of daos we populated on the blockchain during genesis by looking for InitialSchemesSet events
+    // Also load all the proposals for every DAO
+    const newDaoEventsWatcher = daoCreator.InitialSchemesSet({}, { fromBlock: lastBlock });
+    const getNewDaoEvents = promisify(newDaoEventsWatcher.get.bind(newDaoEventsWatcher));
+    const newDaoEvents = await getNewDaoEvents();
+    for (let index = 0; index < newDaoEvents.length; index++) {
+      const event = newDaoEvents[index];
+      daos[event.args._avatar] = await arcActions.getDAOData(event.args._avatar, true);
+    }
+  } catch (e) {
+    console.error("Error getting DAOs: ", e);
+    return;
+  }
+  initialState.daosLoaded = true;
+
+  const normalizedData = normalize(daos, schemas.daoList);
+  initialState.daos = normalizedData.entities.daos || {};
+  initialState.proposals = normalizedData.entities.proposals || {};
+
+  const genesisProtocol = await Arc.GenesisProtocol.deployed();
+
+  let proposalDetails;
+  try {
+    // Get all votes ever made
+    const voteEventWatcher = genesisProtocol.VoteProposal({}, { fromBlock: lastBlock });
+    const getVoteEvents = promisify(voteEventWatcher.get.bind(voteEventWatcher));
+    const voteEvents = await getVoteEvents();
+    let voteEventArgs;
+    for (let cnt = 0; cnt < voteEvents.length; cnt++) {
+      voteEventArgs = voteEvents[cnt].args;
+      proposalDetails = initialState.proposals[voteEventArgs._proposalId];
+
+      // If voter not already added as a member on the DAO set them up (this should not happen right?)
+      if (!initialState.daos[proposalDetails.daoAvatarAddress].members[voteEventArgs._voter]) {
```

Comment:
how/why would it do that? we want to import the data as quickly and easily as possible. remember that everything else is running on the client. this is running on the server

Reply:
oh I see, you are saying create the state object through the reducers. hmm. that might make sense... seems like extra unnecessary work right now :) ill have to ponder if that makes sense as a refactor

Reply:
This can be brittle if we change stuff in the app. If we use the already defined reducers, there can't be inconsistencies between the app's state and this script's state. In terms of performance, there's no difference.